pkg/controllers/user/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -0,0 +1,352 @@
+	multiClusterAppIDSelector = "mcapp"
+	projectIDFieldLabel       = "field.cattle.io/projectId"
+	multiClusterAppPrefix     = "mcapp-"
+	globalNamespace           = "cattle-global-data"
+)
+
+type MCAppController struct {
+	apps                  pv3.AppInterface
+	appLister             pv3.AppLister
+	multiClusterApps      v3.MultiClusterAppInterface
+	multiClusterAppLister v3.MultiClusterAppLister
+	namespaces            corev1.NamespaceInterface
+	templateVersionLister v3.TemplateVersionLister
+	clusterLister         v3.ClusterLister
+	projectLister         v3.ProjectLister
+	clusterName           string
+}
+
+func Register(ctx context.Context, cluster *config.UserContext) {
+	m := &MCAppController{
+		apps:                  cluster.Management.Project.Apps(""),
+		appLister:             cluster.Management.Project.Apps("").Controller().Lister(),
+		namespaces:            cluster.Core.Namespaces(""),
+		multiClusterApps:      cluster.Management.Management.MultiClusterApps(""),
+		multiClusterAppLister: cluster.Management.Management.MultiClusterApps("").Controller().Lister(),
+		clusterLister:         cluster.Management.Management.Clusters("").Controller().Lister(),
+		projectLister:         cluster.Management.Management.Projects("").Controller().Lister(),
+		clusterName:           cluster.ClusterName,
+		templateVersionLister: cluster.Management.Management.TemplateVersions("").Controller().Lister(),
+	}
+	m.multiClusterApps.AddHandler(ctx, "multi-cluster-app-controller", m.sync)
+}
+
+func (m *MCAppController) sync(key string, mcapp *v3.MultiClusterApp) (runtime.Object, error) {
+	if mcapp == nil || mcapp.DeletionTimestamp != nil {
+		_, mcappName, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			return mcapp, err
+		}
+		return m.deleteApps(mcappName, mcapp)
+	}
+	metaAccessor, err := meta.Accessor(mcapp)
+	if err != nil {
+		return mcapp, err
+	}
+	creatorID, ok := metaAccessor.GetAnnotations()[creatorIDAnn]
+	if !ok {
+		return mcapp, fmt.Errorf("MultiClusterApp %v has no creatorId annotation. Cannot create apps for %v", metaAccessor.GetName(), mcapp.Name)
+	}
+
+	answerMap := m.createAnswerMap(mcapp.Spec.Answers)
+
+	externalID, mcapp, err := m.getExternalID(mcapp)
+	if err != nil {
+		return mcapp, err
+	}
+
+	appName, mcapp, err := m.getAppName(mcapp, externalID)
+	if err != nil {
+		return mcapp, err
+	}
+
+	// for all targets, create the App{} instance, so that helm controller App lifecycle can pick it up
+	// only one app per project named mcapp-{{mcapp.Name}}
+	var mcappToUpdate *v3.MultiClusterApp
+	appNameInProject := multiClusterAppPrefix + appName
+	ann := make(map[string]string)
+	ann[creatorIDAnn] = creatorID
+	set := labels.Set(map[string]string{multiClusterAppIDSelector: mcapp.Name})
+	for ind, t := range mcapp.Spec.Targets {
+		split := strings.SplitN(t.ProjectName, ":", 2)
+		if len(split) != 2 {
+			return mcapp, fmt.Errorf("error in splitting project ID")
```

Comment:
If we hit these errors, do these errors indicate enough information to know which controller faced error and which multiclusterapp creation has failed?
If not we should add helper statements, add multiclusterapp ID/projectId having problems